solidity/app/contracts/Router.sol

```diff
@@ -114,38 +117,140 @@ abstract contract Router is AbacusConnectionClient, IMessageRecipient {
+        _outbox.checkpoint();
+    }
+
+    /**
+     * @notice Dispatches a message to an enrolled router via the local router's Outbox
+     * and pays interchain gas for the dispatched message.
+     * @notice Does not create a checkpoint on the Outbox.
+     * @dev Reverts if there is no enrolled router for _destination.
+     * @param _destination The domain of the chain to which to send the message.
+     * @param _msg The message to dispatch.
+     * @param _gasPayment The amount of native tokens to pay the Interchain Gas
+     * Paymaster to process the dispatched message.
+     */
+    function _dispatchWithGas(
+        uint32 _destination,
+        bytes memory _msg,
+        uint256 _gasPayment
+    ) internal {
+        // Gets the abacusConnectionManager from storage once to avoid multiple reads.
+        IAbacusConnectionManager _abacusConnectionManager = abacusConnectionManager;
+        _dispatchWithGas(
+            _abacusConnectionManager.outbox(),
+            _abacusConnectionManager.interchainGasPaymaster(),
+            _destination,
+            _msg,
+            _gasPayment
+        );
     }
 
     /**
-     * @notice Dispatches a message to an enrolled router via the local router's
-     * Outbox and pays for message processing on the destination chain.
+     * @notice Dispatches a message to an enrolled router via the local router's Outbox,
+     * pays interchain gas for the dispatched message, and creates a checkpoint.
      * @dev Reverts if there is no enrolled router for _destination.
      * @param _destination The domain of the chain to which to send the message.
      * @param _msg The message to dispatch.
      * @param _gasPayment The amount of native tokens to pay the Interchain Gas
      * Paymaster to process the dispatched message.
      */
-    function _dispatchToRemoteRouterWithGas(
+    function _dispatchWithGasAndCheckpoint(
         uint32 _destination,
         bytes memory _msg,
         uint256 _gasPayment
     ) internal {
-        uint256 leafIndex = _dispatchToRemoteRouter(_destination, _msg);
+        // Gets the abacusConnectionManager and outbox once to avoid multiple storage reads
+        // and calls.
+        IAbacusConnectionManager _abacusConnectionManager = abacusConnectionManager;
+        IOutbox _outbox = _abacusConnectionManager.outbox();
+        _dispatchWithGas(
+            _outbox,
+            _abacusConnectionManager.interchainGasPaymaster(),
+            _destination,
+            _msg,
+            _gasPayment
+        );
+        _outbox.checkpoint();
+    }
+
+    /**
+     * @notice Creates a checkpoint on the local router's Outbox.
+     * @dev If dispatching a single message and immediately checkpointing,
+     * `_dispatchAndCheckpoint` or `_dispatchWithGasAndCheckpoint` should be preferred,
+     * as they will consume less gas than calling `_dispatch` and this function.
+     */
+    function _checkpoint() internal {
```

Comment:
Similar to this convenience function that we don't actually use, we may want to consider a "raw" dispatch that just sends a message to a specific recipient without the router lookup?

Reply:
I could see it being useful for some use cases but given we expect the vast majority of use cases to not require that and that a developer could still call `_outbox().dispatch` directly, I think it may be better to wait until we feel people need/want this?
If we do add this then there's an argument to also have fns like `dispatchDirectlyWithGas`, `dispatchDirectlyAndCheckpoint`, `dispatchDirectlyWithGasAndCheckpoint`, which I think is a little bloating

Reply:
Fair